launcher/src/main/java/org/apache/spark/launcher/LauncherServer.java

```diff
@@ -210,8 +209,13 @@ int getPort() {
    * Removes the client handle from the pending list (in case it's still there), and unrefs
    * the server.
    */
-  void unregister(ChildProcAppHandle handle) {
-    pending.remove(handle.getSecret());
+  void unregister(AbstractAppHandle handle) {
+    for (Map.Entry<String, AbstractAppHandle> e : pending.entrySet()) {
```

Comment:
you could add a system.identityHashCode "secret" to the InProcessAppHandle to keep the old version, though this is fine too

Reply:
nevermind, stupid idea

Reply:
ok one more try --

you are generating a secret whether its in-process or a child process, so why not store that secret in AbstractAppHandle?  this isn't really a big deal, the efficiency difference doesn't matter, but the one line `pending.remove(handle.getSecret())` is also easier to follow.

also could you rename `pending` to `secretToPendingApps` so its a more clear what the key is.

Reply:
I think I had something like that at some point, but it was more code than the current version... there's a little bit of a chicken & egg problem between handles and secrets, and keeping them separate simplified things a bit at least for me.

I'll do the rename.